cocotb/share/lib/embed/gpi_embed.cpp

```diff
@@ -188,172 +188,136 @@ extern "C" COCOTB_EXPORT void _embed_sim_cleanup(void) {
     }
 }
 
-/**
- * @name    Initialization
- * @brief   Called by the simulator on initialization. Load cocotb Python module
- * @ingroup python_c_api
- *
- * GILState before calling: Not held
- *
- * GILState after calling: Not held
- *
- * Makes one call to PyGILState_Ensure and one call to PyGILState_Release
- *
- * Loads the Python module called cocotb and calls the _initialise_testbench
- * function
- */
+namespace {
 
-static int get_module_ref(const char *modname, PyObject **mod) {
-    PyObject *pModule = PyImport_ImportModule(modname);
+template <typename F>
+class Deferable {
```

Comment:
This is basically `gsl::final_action`

https://github.com/microsoft/GSL/blob/84aeb59f269ee9747e85d780c8aa180bca3a9a60/include/gsl/util#L50-L89

which is basically [E.19 from the C++ code guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-finally).

I'm not saying we should depend on the gsl, but I think it would be a good idea to reuse the name `final_action`.

Any chance you could put this `Deferable` cleanup in a small standalone PR?

Reply:
The whole point of making it was so I didn't have to think through all the object lifetime issues in sim_embed_init and which things need to be freed at which exit points. If I split it out, I did the work I was trying to avoid and it's pointless. If not introducing it now is preferable I'll just refactor it out.

Reply:
It was also accepted into the C++20 standard https://en.cppreference.com/w/cpp/experimental/scope_exit.